### PR TITLE
Make MechanismSimulation an Interface + WPIMechanismSimulation -> core

### PR DESCRIPTION
**Pull request by drorta**

### CODE REVIEW

File: src/main/java/frc/robot/hardware/mechanisms/SimpleMotorSimulation.java [hunk 1]
```diff
@@ -4,13 +4,29 @@
...
 
 	public SimpleMotorSimulation(DCMotorSim motorSimulation, double gearRatio) {
-		super(gearRatio);
 		this.motorSimulation = motorSimulation;
+		this.gearRatio = gearRatio;
+	}
+
+	@Override
+	public double getGearRatio() {
+		return gearRatio;
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/mechanisms/SimpleMotorSimulation.java:19]: DCMotorSim has a gear ratio getter

**Review by Nitay4 — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java [hunk 2]
```diff
@@ -2,29 +2,21 @@
...
 
-	MechanismSimulation(double gearRatio) {
-		this.gearRatio = gearRatio;
-	}
+	Rotation2d getRotorPosition();
 
-	public Rotation2d getRotorPosition() {
-		return getSystemPosition().times(gearRatio);
-	}
+	Rotation2d getRotorVelocityAnglesPerSecond();
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java:11]: angles per seconds? Maybe rotation2d per seconds. Angles are not units, but seconds are. It physically doesn't work. 

Also, angles should be in singular and not plural. @YHerm

File: src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java [hunk 3]
```diff
@@ -2,29 +2,21 @@
...
-		return getSystemVelocityAnglesPerSecond().times(gearRatio);
-	}
 
+	Rotation2d getSystemPosition();
 
-	public abstract Rotation2d getSystemPosition();
+	Rotation2d getSystemVelocityAnglesPerSecond();
 
-	public abstract Rotation2d getSystemVelocityAnglesPerSecond();
+	void setInputVoltage(double voltage);
```
Comment (Sh-Pe) [inline src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java:18]: this just doesn't make sense. If we need to specify that velocity are in Rotation2d/sec, why isn't it VoltageVolt? 

ik this problematic naming occurs outside to this PR. But I wouldn't want to approve inconsistent code.

**Review by hillelv — CHANGES_REQUESTED** [review]
do better

File: src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java [hunk 4]
```diff
@@ -2,29 +2,21 @@
...
 
-	MechanismSimulation(double gearRatio) {
-		this.gearRatio = gearRatio;
-	}
+	Rotation2d getRotorPosition();
 
-	public Rotation2d getRotorPosition() {
-		return getSystemPosition().times(gearRatio);
-	}
+	Rotation2d getRotorVelocityAnglesPerSecond();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java:12]: remove space

File: src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java [hunk 5]
```diff
@@ -2,29 +2,21 @@
 
 import edu.wpi.first.math.geometry.Rotation2d;
 
-public abstract class MechanismSimulation {
+public interface MechanismSimulation {
 
-	private final double gearRatio;
+	double getGearRatio();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java:7]: why?

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java (hunk 5 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/mechanisms/MechanismSimulation.java:7]: *why u need the function

File: src/main/java/frc/robot/hardware/mechanisms/WPIMechanismSimulation.java
Comment (YHerm) [inline src/main/java/frc/robot/hardware/mechanisms/WPIMechanismSimulation.java:1]: put in package with the simple

File: src/main/java/frc/robot/hardware/mechanisms/WPIMechanismSimulation.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package frc.robot.hardware.mechanisms;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public abstract class WPIMechanismSimulation implements MechanismSimulation {
+
+	private final double gearRatio;
+
+	WPIMechanismSimulation(double gearRatio) {
+		this.gearRatio = gearRatio;
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/mechanisms/WPIMechanismSimulation.java:10]: why not make it a function like raz wanted?

File: src/main/java/frc/robot/hardware/mechanisms/wpi/WPIMechanismSimulation.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	@Override
+	public Rotation2d getRotorPosition() {
+		return getSystemPosition().times(getGearRatio());
+	}
+
+	@Override
+	public Rotation2d getRotorVelocityAnglesPerSecond() {
+		return getSystemVelocityAnglesPerSecond().times(getGearRatio());
+	}
+
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/mechanisms/wpi/WPIMechanismSimulation.java:17]: space

**Review by YoniKiriaty — DISMISSED** [review]

File: src/main/java/frc/robot/hardware/mechanisms/wpi/WPIMechanismSimulation.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package frc.robot.hardware.mechanisms.wpi;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.mechanisms.MechanismSimulation;
+
+public abstract class WPIMechanismSimulation implements MechanismSimulation {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/mechanisms/wpi/WPIMechanismSimulation.java:6]: make it interface with default function

**Review by hillelv — CHANGES_REQUESTED** [review]
wtf is wpi mechanism????

**Review by bossyshrimp68 — APPROVED** [review]

**Review by YoniKiriaty — APPROVED** [review]

**Review by maya1414 — APPROVED** [review]

**Review by hillelv — APPROVED** [review]

**Review by YHerm — APPROVED** [review]

**Pull request merged**